middleware/limiter/limiter.go

```diff
@@ -55,34 +55,48 @@ func New(config ...Config) fiber.Handler {
 
 		// Lock entry
 		mux.Lock()
-
 		// Get entry from pool and release when finished
 		e := manager.get(key)
 
 		// Get timestamp
 		ts := atomic.LoadUint64(&timestamp)
-
 		// Set expiration if entry does not exist
 		if e.exp == 0 {
 			e.exp = ts + expiration
 
 		} else if ts >= e.exp {
 			// Check if entry is expired
-			e.hits = 0
-			e.exp = ts + expiration
+			e.prevHits = e.currHits
+			e.currHits = 0
+
+			// checks how into the current window it is and sets the
+			// expiry based on that, otherwise this would only reset on
+			// the next request and not show the correct expiry
+			elapsed := ts - e.exp
+			if elapsed >= expiration {
+				e.exp = ts + expiration
+			} else {
+				e.exp = ts + expiration - elapsed
+			}
 		}
 
 		// Increment hits
-		e.hits++
+		e.currHits++
 
 		// Calculate when it resets in seconds
 		expire := e.exp - ts
 
-		// Set how many hits we have left
-		remaining := cfg.Max - e.hits
+		// weight = time elapsed in current window / total window length
+		weight := float64(expiration-expire) / float64(expiration)
+
+		// rate = request count in previous window - weight + request count in current window
+		rate := int(float64(e.prevHits)*weight) + e.currHits
+
+		// Calculate how many hits can be made based on the current rate
+		remaining := cfg.Max - rate
 
-		// Update storage
-		manager.set(key, e, cfg.Expiration)
+		// Update storage. Garbage collect after 2 windows
+		manager.set(key, e, cfg.Expiration*2)
```

Comment:
I don't think it is a good idea to use 2 windows. Can't you implement this with 1 window?

Reply:
We need to store the key for the previous window and the current window at minimum.